File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
#최단거리 bfs
#경로에 대한 정보 저장 dfs
from collections import deque
def solution(maps):
    dx = [1,-1,0,0]
    dy = [0,0,1,-1]
    
    n = len(maps)
    m = len(maps[0])
    
    start_x, start_y = 0 , 0
    end_x, end_y = n-1,m-1
    
    q = deque()
    #q의 원소는 x좌표, y좌표, 거리
    q.append((start_x, start_y,1))

    visited = set()
    visited.add((start_x, start_y))
    answer = -1
    while q:
        
        x,y,distance = q.popleft()
        
        if(x,y) == (end_x,end_y):
            answer = distance
            
        for d in range(4):
            nx = x + dx[d]
            ny = y + dy[d]
            
            if 0<=nx<n and 0<=ny<m and maps[nx][ny] == 1 and (nx,ny) not in visited:
                q.append((nx,ny, distance+1))
                visited.add((nx,ny))
                 
                
                
    return answer
            

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.java
import java.util.*;

class Solution {
    public int solution(int[][] maps) {
        int n = maps.length;
        int m = maps[0].length;
        
        Deque<int[]> queue = new ArrayDeque<>();
        boolean[][] visited = new boolean[n][m];
        
        queue.offer(new int[] {0,0,1});
        visited[0][0] = true;
        
        while(!queue.isEmpty()){
            int[] cur = queue.poll();
            int x = cur[0];
            int y = cur[1];
            int dist = cur[2];
            
            
            int[] dx = {-1,0,1,0};
            int[] dy = {0,1,0,-1};
            
            for(int i =0;i<4;i++){
                int nx = dx[i] + x;
                int ny = dy[i] + y;
                
                if(nx == n-1 && ny == m-1) return dist+1;
                
                if(nx<0 || ny <0 || nx>=n || ny>=m || maps[nx][ny]==0) continue;
                
                if(visited[nx][ny]) continue;
                
                queue.offer(new int[] {nx,ny,dist+1});
                visited[nx][ny] = true;
                
            }
        }
        return -1;
    }
}